### PR TITLE
KAFKA-14563: add feature version supports in the producers and ApiVersions.   (#117)

### PR DESCRIPTION
**Pull request by CalvinLiu7947**
Add the client side feature version handling.

### CODE REVIEW

**Review by artemlivshits — APPROVED** [review]
LGTM

**Comment by jolshan** [comment]
Minor nit, but can we change the title to be more reflective that this is a change to the producer and ApiVersions?

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java [hunk 1]
```diff
@@ -148,28 +150,36 @@ public class TransactionManagerTest {
...
-    private void initializeTransactionManager(Optional<String> transactionalId) {
+    private void initializeTransactionManager(Optional<String> transactionalId, boolean transactionV2Enabled) {
         Metrics metrics = new Metrics(time);
 
-        apiVersions.update("0", NodeApiVersions.create(Arrays.asList(
-                new ApiVersion()
-                    .setApiKey(ApiKeys.INIT_PRODUCER_ID.id)
-                    .setMinVersion((short) 0)
-                    .setMaxVersion((short) 3),
-                new ApiVersion()
```
Comment (jolshan) [inline clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java:169]: Did we mean to remove produce here?

File: clients/src/test/java/org/apache/kafka/clients/ApiVersionsTest.java [hunk 2]
```diff
@@ -56,4 +57,44 @@ public void testMaxUsableProduceMagicWithRaftController() {
                 .setMaxVersion((short) 2))));
         assertEquals(RecordBatch.CURRENT_MAGIC_VALUE, apiVersions.maxUsableProduceMagic());
     }
+
```
Comment (jolshan) [inline clients/src/test/java/org/apache/kafka/clients/ApiVersionsTest.java:60]: nit: extra new line here

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java (hunk 1 above)
Comment (CalvinLiu7947) [reply clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java:169]: Good catch. Surprisingly it does not break anything.

**Review by jolshan — APPROVED** [review]
Let's just update the title

**Pull request merged**